src/goto-programs/remove_exceptions.cpp

```diff
@@ -98,175 +98,114 @@ class remove_exceptionst
-    }
-
-    bool assertion_error=false;
     if(instr_it->is_throw())
     {
       const exprt &exc =
         uncaught_exceptions_domaint::get_exception_symbol(instr_it->code);
-      assertion_error =
+      bool is_assertion_error =
         id2string(uncaught_exceptions_domaint::get_exception_type(exc.type())).
         find("java.lang.AssertionError")!=std::string::npos;
+      if(!is_assertion_error)
+        return true;
     }
 
-    // if we find a throw different from AssertionError or a function call
-    // that may escape exceptions, then we add an exceptional return
-    // variable
-    if((instr_it->is_throw() && !assertion_error)
-       || has_uncaught_exceptions)
+    if(instr_it->is_function_call())
     {
-      // look up the function symbol
-      symbol_tablet::symbolst::const_iterator s_it=
-        symbol_table.symbols.find(function_id);
-
-      INVARIANT(
-        s_it!=symbol_table.symbols.end(),
-        "functions should be recorded in the symbol table");
-
-      auxiliary_symbolt new_symbol;
-      new_symbol.is_static_lifetime=true;
-      new_symbol.module=function_symbol.module;
-      new_symbol.base_name=id2string(function_symbol.base_name)+EXC_SUFFIX;
-      new_symbol.name=id2string(function_symbol.name)+EXC_SUFFIX;
-      new_symbol.mode=function_symbol.mode;
-      new_symbol.type=pointer_type(empty_typet());
-      symbol_table.add(new_symbol);
-
-      // initialize the exceptional return with NULL
-      symbol_exprt lhs_expr_null=new_symbol.symbol_expr();
-      null_pointer_exprt rhs_expr_null(pointer_type(empty_typet()));
-      goto_programt::targett t_null=
-        goto_program.insert_before(goto_program.instructions.begin());
-      t_null->make_assignment();
-      t_null->source_location=
-        goto_program.instructions.begin()->source_location;
-      t_null->code=code_assignt(
-        lhs_expr_null,
-        rhs_expr_null);
-      t_null->function=function_id;
-
-      break;
+      const exprt &function_expr=
+        to_code_function_call(instr_it->code).function();
+      DATA_INVARIANT(
+        function_expr.id()==ID_symbol,
+        "identifier expected to be a symbol");
+      const irep_idt &function_name=
+        to_symbol_expr(function_expr).get_identifier();
+      if(!exceptions_map[function_name].empty())
+        return true;
     }
   }
+
+  return false;
 }
 
 /// Translates an exception landing-pad into instructions that copy the
 /// in-flight exception pointer to a nominated expression, then clear the
 /// in-flight exception (i.e. null the pointer), hence marking it caught.
-/// \param function_id: name of the function containing this landingpad
-///   instruction
 /// \param goto_program: body of the function containing this landingpad
 ///   instruction
 /// \param instr_it: iterator pointing to the landingpad instruction.
 ///   Will be overwritten.
+/// \param may_catch: if true, an exception may be caught here; otherwise
+///   the catch site is unreachable.
```

Comment:
Consider explaining in this comment when this would be `true` and when it would be `false`